go/vt/topotools/rebuild_keyspace.go

```diff
@@ -78,39 +57,66 @@ func RebuildKeyspaceLocked(ctx context.Context, log logutil.Logger, ts *topo.Ser
 		return err
 	}
 
+	// The caller intents to update all cells in this case
+	if len(cells) == 0 {
+		cells, err = ts.GetCellInfoNames(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
 	shards, err := ts.FindAllShardsInKeyspace(ctx, keyspace)
 	if err != nil {
 		return err
 	}
 
+	// This is safe to rebuild as long there are not srvKeyspaces with tablet controls set.
+	// TODO: Add this validation.
+
 	// Build the list of cells to work on: we get the union
 	// of all the Cells of all the Shards, limited to the provided cells.
 	//
 	// srvKeyspaceMap is a map:
 	//   key: cell
 	//   value: topo.SrvKeyspace object being built
 	srvKeyspaceMap := make(map[string]*topodatapb.SrvKeyspace)
-	findCellsForRebuild(ki, shards, cells, srvKeyspaceMap)
+	for _, cell := range cells {
+		srvKeyspace, err := ts.GetSrvKeyspace(ctx, cell, keyspace)
+		switch {
+		case err == nil:
+			for _, partition := range srvKeyspace.GetPartitions() {
+				if partition.GetShardTabletControls() != nil {
+					return fmt.Errorf("can't rebuild serving keyspace while a migration is on going. TabletControls is set for partition %v", partition)
+				}
 
-	// Then we add the cells from the keyspaces we might be 'ServedFrom'.
-	for _, ksf := range ki.ServedFroms {
-		servedFromShards, err := ts.FindAllShardsInKeyspace(ctx, ksf.Keyspace)
-		if err != nil {
+			}
+		case topo.IsErrType(err, topo.NoNode):
+			// NOOP
+		default:
 			return err
 		}
-		findCellsForRebuild(ki, servedFromShards, cells, srvKeyspaceMap)
+		srvKeyspaceMap[cell] = &topodatapb.SrvKeyspace{
+			ShardingColumnName: ki.ShardingColumnName,
+			ShardingColumnType: ki.ShardingColumnType,
+			ServedFrom:         ki.ComputeCellServedFrom(cell),
+		}
 	}
 
+	servedTypes := []topodatapb.TabletType{topodatapb.TabletType_MASTER, topodatapb.TabletType_REPLICA, topodatapb.TabletType_RDONLY}
+
 	// for each entry in the srvKeyspaceMap map, we do the following:
 	// - get the Shard structures for each shard / cell
 	// - if not present, build an empty one from global Shard
-	// - compute the union of the db types (replica, master, ...)
 	// - sort the shards in the list by range
 	// - check the ranges are compatible (no hole, covers everything)
 	for cell, srvKeyspace := range srvKeyspaceMap {
 		for _, si := range shards {
-			servedTypes := si.GetServedTypesPerCell(cell)
-
+			// We rebuild keyspace iff:
+			// 1) shard master is in a serving state.
+			// 2) shard has served type for master (this is for backwards compatibility).
+			if !(si.IsMasterServing || si.GetServedType(topodatapb.TabletType_MASTER) != nil) {
```

Comment:
* This is one of the core parts of this change. @sougou  and I realized that if a master is serving, from a logical perspective that's the only thing you need to know to generate the SrvKeyspaceGraph. You could be in one of these scenarios:

1) User creates a new keyspace with new shards. In init tablet, because there are no overlapping shards IsMasterServing will be set to true. When you build the keyspace graph for this newly created keyspace, this code will generate the serving graph for all types: **rdonly, replica, master**. 
2) If you are creating shards in an existent keyspaces (to prepare for a split), InitTablet will IsMasterServing to false, and rebuild keyspace won't take those new shards into consideration for any type. 

When you are in the process of migrating serving types, there is code there that will modify the serving keyspace graph to add/remove new shards accordingly. But that does not go through this code path.